File: src/leetcode/second/Code_144_PreorderTraversal.java
package leetcode.second;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import tool.TreeNode;

/**
 * 二叉树的先序遍历
 * 迭代方式
 * 
 * @since 2020-8-4 Tuesday 8:54 - 9:10
 */
public class Code_144_PreorderTraversal {
    static List<Integer> preorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();
        if (root != null) {
            stack.push(root);
        }
        while (!stack.isEmpty()) {
            if (root == null) {
                res.add(stack.pop().val);
                continue;
            }
            res.add(stack.pop().val);
            if (root.right != null) {
                stack.push(root.right);
            }
            if (root.left != null) {
                stack.push(root.left);
            }
            if (root.left != null) {
                root = root.left;
            } else if (root.right != null) {
                root = root.right;
            } else {
                if (!stack.isEmpty()) {
                    root = stack.peek();
                } else {
                    root = null;
                }
            }
        }
        return res;
    }

    static List<Integer> _preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root != null) {
            Stack<TreeNode> stack = new Stack<>();
            stack.push(root);
            while (!stack.isEmpty()) {
                root = stack.pop();
                res.add(root.val);
                if (root.right != null) {
                    stack.push(root.right);
                }
                if (root.left != null) {
                    stack.push(root.left);
                }
            }
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(preorderTraversal(TreeNode.tree())); // 1, 2, 4, 5, 3, 6, 7
        System.out.println(_preorderTraversal(TreeNode.tree())); // 1, 2, 4, 5, 3, 6, 7
    }
}

File: src/cpp-leetcode/first/Code_144_PreorderTraversal.cpp
#include <iostream>
#include <stack>
#include <vector>

#include "../Tree.h"

/**
 * binary tree preorder traversal
 *
 * @since 2020-9-14 Monday 14:43 - 14:50
 */
class Solution {
   public:
    std::vector<int> preorderTraversal(TreeNode* root) {
        if (root == NULL)
            return {};
        std::vector<int> res;
        std::stack<TreeNode*> stack;
        stack.push(root);
        while (!stack.empty()) {
            TreeNode* cur = stack.top();
            stack.pop();
            res.push_back(cur->val);
            if (cur->right != NULL) {
                stack.push(cur->right);
            }
            if (cur->left != NULL) {
                stack.push(cur->left);
            }
        }
        return res;
    }
};

int main(int argc, char const* argv[]) {
    TreeNode* root = Tree::binaryTree();
    Solution s;
    std::vector<int> res = s.preorderTraversal(root);
    for (int i : res) {
        std::cout << i << std::endl;
    }
}
